[DNM]VXFM-6201 :Verify the SVM we are configuring is on the correct host

pankaj1176088 opened the pull request:
@gavin-scott 
Please review the approach .

gavin-scott commented on lib/puppet/provider/vc_vm/default.rb:
```diff
@@ -1462,7 +1462,11 @@ def datacenter(name=resource[:datacenter])
+      host ||= cluster.host.select{ |h| h[:name] == resource[:host] }.first if cluster && cluster.host
+      @vm ||= host.vm.select{ |v| v[:name] == resource[:name]}.first if host && host.vm
+    end
+    @vm
```
Hmm, I'm not sure about this. `findvm_by_name` has some fairly complicated different logic it can do that you're losing:

https://github.com/dell-asm/vmware-vcenter/blob/1fbc59ed736a2a075efed6eb9be93d3914796b2d/lib/puppet/provider/vcenter.rb#L63-L81

I'd have to dig into this more to find the right fix. @sushilrai do you have any ideas?

pankaj1176088 commented:
@gavin-scott  @sushilrai 
In existing deployment flow all SVM is created inside the vmFolder . If we deploy the another cluster within same datacenter then SVM is not being created because of duplicate SVM name . Since both are pointing to vmFolder . 

Instead of pointing to vmFolder . Here we added sub folder with cluster name . And Now SVM is created inside the subfolder . 
        dc
         |
    vmFolder
|                     |
Cluster1      Cluster2
-SVM1        -SVM1
-SVM2        -SVM2
-SVM3        -SVM3

Please review the approach .

pankaj1176088 commented:
@gavin-scott 

In the Existing code under the vmware module  "exists?" method only check SVM name .  It return true if vm has same name even though vm does not belong to that cluster . 

To Fixed this issue we added condition to check cluster name as well as .

Like vm.name == vm_name && vm.resourcePool.owner.name == cluster_name . 

So now it continues with flow . But it failed during the OVF deployment . 

Debug: Failure during OVF deployment for vm: svm-node3 with error DuplicateName: The name 'svm-node3' already exists.: RbVmomi::Fault

This is the api it calls for ovf deployment .

https://github.com/dell-asm/vmware-vcenter/blob/5c078189dbc5196ba9ce5ed3aea08c35065abbf0/lib/puppet/provider/vc_vm/default.rb#L1319

It failed because both point to vmFolder . 

To avoid duplicate name , we added sub folder with cluster name . And  SVM is getting created inside the sub folder . This approach is working fine to create two svm with same name under the same data center .



Could you please let us know if any other approach we can use .
